vllm/sequence.py

```diff
@@ -122,16 +122,42 @@ def __init__(
         if output_token_ids is None:
             output_token_ids = []
 
-        self.prompt_token_ids = prompt_token_ids
-        self._prompt_token_ids_tuple = tuple(prompt_token_ids)
-        self.output_token_ids = output_token_ids
+        self._prompt_token_ids = prompt_token_ids
+        self._prompt_token_ids_tuple: Tuple[int, ...] = tuple(prompt_token_ids)
+        self._output_token_ids = output_token_ids
         self.cumulative_logprob = 0.0
         # The number of tokens that are computed (that run against the model).
         self._num_computed_tokens = 0
         self._stage: SequenceStage = SequenceStage.PREFILL
 
+        self._update_cached_all_tokens()
+
+    def _update_cached_all_tokens(self):
+        self._cached_all_token_ids = (self._prompt_token_ids +
+                                      self._output_token_ids)
+
+    @property
+    def prompt_token_ids(self) -> List[int]:
+        return self._prompt_token_ids
```

Comment:
I think we should return a tuple/shallow copy so that this and also output_token_ids doesn't get modified by mistake (and thus bypass `_update_cached_all_tokens`)

Reply:
Good catch, changed the return types to be tuples.

Reply:
I have changed the approach here to protect accesses to prompt_token_ids and output_token_ids. Now, it uses a class MonitoredList that records a timestamp of the last update, and based on that, the cached all tokens is updated. I did in this way to avoid changing all usages of the prompt/output token ids due to tuple change and also it avoids unnecessary copies of list => tuples which are also expensive.

Reply:
@Yard1 one found out that there is actually an issue with the deserialization with ray, so I have removed this and made the prompt/output token_ids accessors return tuples. It introduces a conversion for the output_token_ids to tuple but it seems not to be bad and the performance is still good. To make it work, I have propagated the tuple type upward in the vllm software stack, since we don't expect seq_data users to use these accessors to change data (but only via the append_token() function)